File: 0277-find-the-celebrity/0277-find-the-celebrity.java
/* The knows API is defined in the parent class Relation.
      boolean knows(int a, int b); */

public class Solution extends Relation {
    public int findCelebrity(int n) {
        
        int i = 0;
        int j = n - 1;
        while (j >= 0 ) {
            if (knows(i, j) && (i == j ||!knows(j, i))) {
                i++;
                if (i >= n) {
                    return j;
                }
            } else {
                j--;
                i = 0;
            }
        }

        return -1;
    }
}

File: 0277-find-the-celebrity/0277-find-the-celebrity.py
# The knows API is already defined for you.
# return a bool, whether a knows b
# def knows(a: int, b: int) -> bool:

class Solution:
    def findCelebrity(self, n: int) -> int:
        canBe = [True] * n
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                
                know = knows(i, j)
                if know:
                    canBe[i] = False
                    break
                else:
                    canBe[j] = False

        for i in range(n):
            if canBe[i]:
                isValid = True
                for j in range(n):
                    if not knows(j, i):
                        isValid = False
                        canBe[j] = False
                        break
                if isValid:

                    return i
        
        return -1
            
